js/modules/k6/experimental/fs/file.go

```diff
@@ -13,7 +14,7 @@ type file struct {
 	data []byte
 
 	// offset holds the current offset in the file
-	offset int
+	offset atomic.Int64
```

Comment:
```suggestion
    // we should explain here that this is a partial solution.
	offset atomic.Int64
```
This is a partial solution and the logic could still have race condition if we run Read+Seek. We should explicit that this is delegated to the user of caring of it and not writing racy logic.
We also should add a suggestion that if this becomes an issue then synchronizing the offset movements with a channel to keep them in order could be a better solution.

Reply:
Yeah, I do feel like the currently proposed solution is okay *but* from my experiments deno [^1] - it actually sequences the reads and seeks (and likely writes) so that if you have 

```javascript
let s1 = file.operation1(...)
let s2 = file.operation1(...)
let s3 = file.operation1(...)
await s3;
```
Actually both s1 and s2 also execute in that order instead of any other.

This likely has no practical benefit, but might be better to just have it this way from the very beginning. This likely can be made through having some kind of list of operations or just each operation referencing the last operation added and waiting for them :shrug: .

In the current implementation for example 
```javascript
let s1 = file.seek(10, 1)
let s2 = file.seek(10, 1)
await s2;
await file.read(...);
// there doesn't seem to be any gurantee whether the offset here will be 10 or 20
// but with deno it is *always* 20 in my experiments.
```

[^1]: This is purely on my experiments with it. No part of the documentation makes this claim or anything against it.

Reply:
I seem to have missed that Deno behavior specifically indeed. It does appear that in this scenario the operations are indeed kept in the input sequence.

Building on what @codebien mentioned in his previous comments, I had explored the idea of passing "operations/commands" messages around, as opposed to immediately executing operations, but as the benefits were not immediately clear, I left it at that. I'm gonna look into it again 👍🏻

Reply:
Again - I am not certain we should or must do the same as Deno - I am just pointing out that they seem to have a specific behaviour. While the currently implemented one not only deviates from theirs, which is not really a problem IMO, and also has arguably a harder to reason behaviour - which IMO is the bigger problem. 

Both of those IMO are less blocking and we can still have the more "racy" behaviour or something different from Deno that still is more "stable". I am not certain what that would be though :shrug: .

>  I had explored the idea of passing "operations/commands" messages around

I have not tried implementing this so no idea how well that will work or not, but it seems a bit complicated with moving parts. Arguably it is aligned with the "share by communicating" slogan of golang.

But to expand on my comment 
> or just each operation referencing the last operation added and waiting for them 🤷 .

Here I mean to have the ... file struct ... for example (:thinking:) have a "last currently not finished operation". All operations when created look at that field and replace it with "themselves". Then if the field had previously value "await" it to finish before they start to execute.

This seems like way easier to implement (from the current perspective of having not tried it). If you do all the manipulation of that field on the eventloop - also drops on the communication perspectice significantly.

I don't think we have a use case for knowing there are N operations or reordering them so having a single-linked-list of sort seems fine. 

But again - I have not implemented something like this in the context of k6 and goja + eventloop, so maybe something else will be nicer :shrug: 

This whole thing is slightly related to https://github.com/grafana/xk6-timers/issues/3 although maybe just in theory as that is about timers and waiting for them.

Reply:
>I'm gonna look into it again 👍🏻

We resolved the main data race, and now this new issue is not a must-have. As @mstoykov reported, there isn't a spec defining it, so developers shouldn't have expectations on this behavior as it can differ between implementations.
However, it's still nice to have for us to align with the rest of the JavaScript ecosystem, so I'm for just opening an issue and setting a TODO comment in the code.

Reply:
Thanks for your input @mstoykov, you actually started answering the main question I had about the implementation of your idea, so cheers for that 🙇🏻 

I will open a dedicated issue and a TODO in the code then 👍🏻

Reply:
See issue #3433